crates/flux-infer/src/refine_tree.rs

```diff
@@ -780,3 +791,328 @@ impl RefineCtxtTrace {
+                        };
+                        assms.remove(&kvid);
+                        if assignment.contains(rhs_kvid) && assms.is_empty() {
+                            candidates.push(cid);
+                        }
+                    };
+                }
+            }
+        }
+
+        assignment
+    }
+
+    /// Computes the set of all kvars that can be assigned to Top (True),
+    /// because they do not (transitively) reach any concrete HEAD.
+    fn top_kvars(&self, graph: Graph) -> Assignment {
+        // initialize
+        let mut assignment = Assignment::new(graph.kvars(), Label::Top);
+
+        // set of kvar {k | cid in graph.edges, c.rhs is concrete, k in c.lhs }
+        let mut candidates = vec![];
+        for info in graph.edges.values() {
+            if matches!(info.rhs, VertexId::Conc) {
+                for kvid in &info.lhs {
+                    candidates.push(*kvid);
+                }
+            }
+        }
+
+        // set each kvar that transitively reaches a concrete HEAD to NON-BOT
+        while let Some(kvid) = candidates.pop() {
+            // set that kvar to non-top
+            assignment.remove(kvid);
+
+            // for each constraint where kvid appears as head
+            for cid in graph.kv_rhs.get(&kvid).unwrap_or(&FxHashSet::default()) {
+                let info = graph
+                    .edges
+                    .get(cid)
+                    .unwrap_or_else(|| tracked_span_bug!("missing constraint info for {:?}", cid));
+                // add kvars in lhs to candidates (if they have not already been solved to non-BOT)
+                for lhs_kvid in &info.lhs {
+                    if assignment.contains(*lhs_kvid) {
+                        candidates.push(*lhs_kvid);
+                    }
+                }
+            }
+        }
+
+        assignment
+    }
+}
+
+#[derive(Debug)]
+struct Graph {
+    /// description of each constraint
+    edges: FxHashMap<CstrId, CstrInfo>,
+    /// set of edges where kvid appears as ASSM
+    kv_lhs: FxHashMap<KVid, FxHashSet<CstrId>>,
+    /// set of edges where kvid appears as HEAD
+    kv_rhs: FxHashMap<KVid, FxHashSet<CstrId>>,
+}
+
+impl Graph {
+    fn new(node: &Node) -> Self {
+        let mut graph = Self {
+            edges: FxHashMap::default(),
+            kv_lhs: FxHashMap::default(),
+            kv_rhs: FxHashMap::default(),
+        };
+        graph.build(node, &mut vec![]);
+        graph
+    }
+
+    fn head(&self, cid: CstrId) -> VertexId {
+        let info = self
+            .edges
+            .get(&cid)
+            .unwrap_or_else(|| tracked_span_bug!("missing constraint info for {:?}", cid));
+        info.rhs
+    }
+
+    fn cstrs_with_kvar_as_assumption(&self, kvid: &KVid) -> Vec<CstrId> {
+        self.kv_lhs
+            .get(kvid)
+            .unwrap_or(&FxHashSet::default())
+            .iter()
+            .copied()
+            .collect()
+    }
```

Comment:
If we are making the values in `kv_lhs` vectors, we can avoid the allocation here with

```suggestion
    fn cstrs_with_kvar_as_assumption(&self, kvid: &KVid) -> &[CstrId] {
        self.kv_lhs.get(kvid).map_or(&[], Vec::as_slice) 
    }
```